hugr-passes/src/dataflow/partial_value.rs

```diff
@@ -0,0 +1,720 @@
+    ///
+    /// # Panics
+    ///
+    /// if the value is believed, for that tag, to have a number of values other than `len`
+    pub fn variant_values(&self, tag: usize, len: usize) -> Option<Vec<PartialValue<V>>> {
+        let vals = match self {
+            PartialValue::Bottom | PartialValue::Value(_) => return None,
+            PartialValue::PartialSum(ps) => ps.variant_values(tag)?,
+            PartialValue::Top => vec![PartialValue::Top; len],
+        };
+        assert_eq!(vals.len(), len);
+        Some(vals)
+    }
+
+    /// Tells us whether this value might be a Sum with the specified `tag`
+    pub fn supports_tag(&self, tag: usize) -> bool {
+        match self {
+            PartialValue::Bottom | PartialValue::Value(_) => false,
+            PartialValue::PartialSum(ps) => ps.supports_tag(tag),
+            PartialValue::Top => true,
+        }
+    }
+
+    /// Turns this instance into some "concrete" value type `V2`, *if* it is a single value,
+    /// or a [Sum](PartialValue::PartialSum) (of a single tag) convertible by
+    /// [PartialSum::try_into_value].
+    ///
+    /// # Errors
+    ///
+    /// If this PartialValue was `Top` or `Bottom`, or was a [PartialSum](PartialValue::PartialSum)
+    /// that could not be converted into a [Sum] by [PartialSum::try_into_value] (e.g. if `typ` is
+    /// incorrect), or if that [Sum] could not be converted into a `V2`.
+    pub fn try_into_value<V2, VE, SE>(self, typ: &Type) -> Result<V2, ExtractValueError<V, VE, SE>>
+    where
+        V2: TryFrom<V, Error = VE> + TryFrom<Sum<V2>, Error = SE>,
+    {
+        match self {
+            Self::Value(v) => V2::try_from(v.clone())
+                .map_err(|e| ExtractValueError::CouldNotConvert(v.clone(), e)),
+            Self::PartialSum(ps) => {
+                let v = ps.try_into_value(typ)?;
+                V2::try_from(v).map_err(ExtractValueError::CouldNotBuildSum)
+            }
+            Self::Top => Err(ExtractValueError::ValueIsTop),
+            Self::Bottom => Err(ExtractValueError::ValueIsBottom),
+        }
+    }
+}
+
+impl TryFrom<Sum<Value>> for Value {
+    type Error = ConstTypeError;
+
+    fn try_from(value: Sum<Value>) -> Result<Self, Self::Error> {
+        Self::sum(value.tag, value.values, value.st)
+    }
+}
+
+impl<V: AbstractValue> Lattice for PartialValue<V> {
+    fn join_mut(&mut self, other: Self) -> bool {
+        self.assert_invariants();
+        match (&*self, other) {
+            (Self::Top, _) => false,
+            (_, other @ Self::Top) => {
+                *self = other;
+                true
+            }
+            (_, Self::Bottom) => false,
+            (Self::Bottom, other) => {
+                *self = other;
+                true
+            }
+            (Self::Value(h1), Self::Value(h2)) => match h1.clone().try_join(h2) {
+                Some(h3) => {
+                    let ch = h3 != *h1;
```

Comment:
Well, first I changed `AbstractValue::try_join` to return `(Self, bool)` i.e. whether it's changed, thus avoiding the comparison. Rather than adding the debug_assert, though, I managed to avoid the `clone` (leaving nothing to compare against) by cunning use of `std::mem::swap`.

A large part of my brain here is saying "premature over-optimization" but...